### PR TITLE
feat: w3c schema endorsement

### PR DESCRIPTION
**Pull request by bhavanakarwade**
### What?
- Implemented w3c schema request endorsement functionality
- Implemented w3c schema submit endorsementfunctionality
- Modified sign transaction functionality according to schema type

### Why?
- To allow and ensure that the system can handle W3C schema requests and submissions properly. 
- Sign transaction functionality is for to correctly differentiates between different schema types to apply appropriate processing logic.

### CODE REVIEW

File: apps/ecosystem/src/ecosystem.service.ts [hunk 1]
```diff
@@ -1557,20 +1663,86 @@ export class EcosystemService {
...
+      const endorsementTransactionPayload = await this.ecosystemRepository.getEndorsementTransactionById(
+        endorsementId,
+        endorsementTransactionStatus.REQUESTED
+      );
+
+      if (!endorsementTransactionPayload) {
+        throw new BadRequestException(ResponseMessages.ecosystem.error.transactionNotRequested);
       }
+
+      const w3cEndorsementTransactionPayload = await this.ecosystemRepository.getEndorsementTransactionByIdAndType(endorsementId, endorsementTransactionType.W3CSCHEMA);
```
Comment (tipusinghaw) [inline apps/ecosystem/src/ecosystem.service.ts:1684]: it's better to destructure and use the response

File: apps/api-gateway/src/ecosystem/ecosystem.service.ts [hunk 2]
```diff
@@ -163,11 +163,12 @@ export class EcosystemService extends BaseService {
   }
 
   async schemaEndorsementRequest(
-    requestSchemaPayload: RequestSchemaDto,
+    requestSchemaPayload: RequestSchemaDto | RequestW3CSchemaDto,
+    schemaType: string,
```
[the remark this answers is not in the document]
Comment (bhavanakarwade) [reply apps/api-gateway/src/ecosystem/ecosystem.service.ts:167]: Resolved

File: apps/ecosystem/enums/ecosystem.enum.ts [hunk 3]
```diff
@@ -23,6 +23,7 @@ export enum endorsementTransactionStatus {
 
 export enum endorsementTransactionType {
     SCHEMA = 'schema',
+    W3CSCHEMA = 'W3C Schema',
```
[the remark this answers is not in the document]
Comment (bhavanakarwade) [reply apps/ecosystem/enums/ecosystem.enum.ts:26]: Resolved

File: apps/ecosystem/interfaces/ecosystem.interfaces.ts [hunk 4]
```diff
@@ -14,6 +19,16 @@ export interface RequestSchemaEndorsement {
...
 }
 
+export interface IRequestW3CSchemaEndorsement {
+  orgId: string;
+  userId?: string;
+  schemaName: string;
+  did: string;
+  schemaAttributes: IW3CSchemaAttributesVAlue[];
+  description: string;
+  endorse?: boolean;
```
[the remark this answers is not in the document]
Comment (bhavanakarwade) [reply apps/ecosystem/interfaces/ecosystem.interfaces.ts:29]: Resolved

File: apps/ecosystem/src/ecosystem.service.ts [hunk 5]
```diff
@@ -801,49 +803,75 @@ export class EcosystemService {
...
   ): Promise<IEndorsementTransaction> {
     try {
-      const { name, version } = requestSchemaPayload;
-      const alreadySchemaExist = await this._schemaExist(version, name);
-      this.logger.log(`alreadySchemaExist ::: ${JSON.stringify(alreadySchemaExist.length)}`);
-
-      if (0 !== alreadySchemaExist.length) {
-        throw new ConflictException(ResponseMessages.ecosystem.error.schemaAlreadyExist);
-      }
+      let alreadySchemaExist;
```
[the remark this answers is not in the document]
Comment (bhavanakarwade) [reply apps/ecosystem/src/ecosystem.service.ts:817]: Resolved

File: apps/ecosystem/src/ecosystem.controller.ts [hunk 6]
```diff
@@ -156,15 +156,14 @@ export class EcosystemController {
    */
   @MessagePattern({ cmd: 'schema-endorsement-request' })
   async schemaEndorsementRequest(payload: {
-    requestSchemaPayload: RequestSchemaEndorsement;
+    requestSchemaPayload: IRequestSchemaEndorsement | IRequestW3CSchemaEndorsement;
+    schemaType: string;
```
[the remark this answers is not in the document]
Comment (bhavanakarwade) [reply apps/ecosystem/src/ecosystem.controller.ts:160]: Resolved

File: apps/api-gateway/src/ecosystem/dtos/request-schema.dto.ts [hunk 7]
```diff
@@ -73,6 +73,66 @@ export class RequestSchemaDto {
...
+  @IsString({ message: 'did must be in string format.' })
+  did: string;
+
+  @ApiProperty()
+  @Transform(({ value }) => trim(value))
+  @IsNotEmpty({ message: 'description is required' })
+  @IsString({ message: 'description must be in string format.' })
+  description: string;
+
+  userId?: string;
```
[the remark this answers is not in the document]
Comment (bhavanakarwade) [reply apps/api-gateway/src/ecosystem/dtos/request-schema.dto.ts:133]: Resolved

File: apps/ecosystem/src/ecosystem.service.ts [hunk 8]
```diff
@@ -801,49 +803,75 @@ export class EcosystemService {
...
         schemaRequestExist,
-        ecosystemMemberDetails,
-        platformConfig,
+        ecosystemOrgAgentDetails,
         ecosystemLeadAgentDetails,
         getEcosystemOrgDetailsByOrgId
       ] = await Promise.all([
         this.ecosystemRepository.findRecordsByNameAndVersion(requestSchemaPayload?.name, requestSchemaPayload?.version),
         this.ecosystemRepository.getAgentDetails(orgId),
-        this.ecosystemRepository.getPlatformConfigDetails(),
```
[the remark this answers is not in the document]
Comment (bhavanakarwade) [reply apps/ecosystem/src/ecosystem.service.ts:846]: Resolved

File: apps/ecosystem/src/ecosystem.service.ts [hunk 9]
```diff
@@ -801,49 +803,75 @@ export class EcosystemService {
...
+  
+        if (isNaN(parseFloat(version)) || version.toString().indexOf('.') === schemaVersionIndexOf) {
+          throw new NotAcceptableException(ResponseMessages.schema.error.invalidVersion);
+        }
+        return await this.requestIndySchemaEndorsement(requestSchemaPayload as IRequestSchemaEndorsement, orgId, ecosystemId);
+      } else if (schemaType === schemaRequestType.W3C) {
+        const { schemaName } = requestSchemaPayload as IRequestW3CSchemaEndorsement;
+        alreadySchemaExist = await this._schemaExist(schemaName);
 
-      const getEcosystemLeadDetails = await this.ecosystemRepository.getEcosystemLeadDetails(ecosystemId);
```
[the remark this answers is not in the document]
Comment (bhavanakarwade) [reply apps/ecosystem/src/ecosystem.service.ts:821]: Initially, we were using only one function for requesting schema endorsement. Now, we have segregated the Indy and W3C schema endorsement functions. I have used getEcosystemLeadDetails for the Indy type schema endorsement request function.

File: apps/ecosystem/src/ecosystem.repository.ts [hunk 10]
```diff
@@ -1009,6 +1009,56 @@ export class EcosystemRepository {
     }
   }
 
+  // eslint-disable-next-line camelcase
+  async getTransactionDetailsByEndorsementId(endorsementId: string): Promise<endorsement_transaction> {
```
[the remark this answers is not in the document]
Comment (bhavanakarwade) [reply apps/ecosystem/src/ecosystem.repository.ts:1013]: In the getEndorsementTransactionById function, we were fetching transaction details using the endorsement ID and status. For my task, I require all details regardless of status. I tried to refactor the getEndorsementTransactionById function to retrieve details for all statuses, but it is used in multiple functions

File: apps/ecosystem/src/ecosystem.service.ts [hunk 11]
```diff
@@ -801,49 +803,75 @@ export class EcosystemService {
...
    *
-   * @param RequestSchemaEndorsement
+   * @param IRequestSchemaEndorsement
    * @returns
    */
+
   async requestSchemaEndorsement(
-    requestSchemaPayload: RequestSchemaEndorsement,
+    requestSchemaPayload: IRequestSchemaEndorsement | IRequestW3CSchemaEndorsement,   
+    schemaType: string,
```
[the remark this answers is not in the document]
Comment (bhavanakarwade) [reply apps/ecosystem/src/ecosystem.service.ts:812]: Resolved

File: apps/ecosystem/src/ecosystem.repository.ts [hunk 12]
```diff
@@ -1019,6 +1069,16 @@ export class EcosystemRepository {
     }
   }
 
+  // eslint-disable-next-line camelcase
+  async findW3CSchemaRecords(schemaName: string): Promise<endorsement_transaction[]> {
```
[the remark this answers is not in the document]
Comment (bhavanakarwade) [reply apps/ecosystem/src/ecosystem.repository.ts:1073]: Resolved

File: apps/api-gateway/src/schema/schema.controller.ts [hunk 13]
```diff
@@ -143,7 +143,7 @@ export class SchemaController {
   @ApiResponse({ status: HttpStatus.CREATED, description: 'Success', type: ApiResponseDto })
   async createW3CSchema(@Res() res: Response, @Body() schemaPayload: CreateW3CSchemaDto,  @Param('orgId', new ParseUUIDPipe({exceptionFactory: (): Error => { throw new BadRequestException(ResponseMessages.organisation.error.invalidOrgId); }})) orgId: string, @User() user: IUserRequestInterface): Promise<Response> {
 
-    const schemaDetails = await this.appService.createW3CSchema(schemaPayload, orgId, user);
+    const schemaDetails = await this.appService.createW3CSchema(schemaPayload, orgId, user.id);
```
[the remark this answers is not in the document]
Comment (bhavanakarwade) [reply apps/api-gateway/src/schema/schema.controller.ts:146]: This function requires only the userId, not the complete user interface.

File: apps/ecosystem/src/ecosystem.service.ts (hunk 1 above)
Comment (bhavanakarwade) [reply apps/ecosystem/src/ecosystem.service.ts:1684]: Resolved

File: apps/ecosystem/src/ecosystem.service.ts [hunk 14]
```diff
@@ -801,49 +803,75 @@ export class EcosystemService {
...
+        }
+  
+        const schemaVersionIndexOf = -1;
+  
+        if (isNaN(parseFloat(version)) || version.toString().indexOf('.') === schemaVersionIndexOf) {
+          throw new NotAcceptableException(ResponseMessages.schema.error.invalidVersion);
+        }
+        return await this.requestIndySchemaEndorsement(requestSchemaPayload as IRequestSchemaEndorsement, orgId, ecosystemId);
+      } else if (schemaType === schemaRequestType.W3C) {
+        const { schemaName } = requestSchemaPayload as IRequestW3CSchemaEndorsement;
```
[the remark this answers is not in the document]
Comment (bhavanakarwade) [reply apps/ecosystem/src/ecosystem.service.ts:842]: Resolved

**Review by KulkarniShashank — APPROVED** [review]
LG

**Review by tipusinghaw — APPROVED** [review]

**Pull request merged**